Libplanet/Assets/Currency.cs

```diff
@@ -61,19 +62,67 @@ namespace Libplanet.Assets
         /// </summary>
         public readonly HashDigest<SHA1> Hash;
 
+        /// <summary>
+        /// Whether the total supply of this instance of <see cref="Currency"/> is trackable.
+        /// </summary>
+        public readonly bool TotalSupplyTrackable;
+
+        private readonly (BigInteger Major, BigInteger Minor)? _maximumSupply;
+
         /// <summary>
         /// Defines a <see cref="Currency"/> type.
         /// </summary>
         /// <param name="ticker">The ticker symbol, e.g., <c>&quot;USD&quot;</c>.</param>
         /// <param name="decimalPlaces">The number of digits to treat as <a
         /// href="https://w.wiki/ZXv#Treatment_of_minor_currency_units_(the_%22exponent%22)">minor
         /// units (i.e., exponent)</a>.</param>
+        /// <param name="maximumSupply">The uppermost quantity of currency allowed to exist.
+        /// See also <see cref="MaximumSupply"/> field which corresponds to this.
+        /// Must be <value>null</value> when <paramref name="totalSupplyTrackable" /> is
+        /// <value>null</value>.</param>
         /// <param name="minters">The <see cref="Address"/>es who can mint the currency.
         /// See also <see cref="Minters"/> field which corresponds to this.</param>
+        /// <param name="totalSupplyTrackable">Whether the total supply of this currency is
+        /// trackable. This parameter is for backwards compatibility with
+        /// <see cref="Currency">Currencies</see> defined before maximum supply was introduced,
+        /// and must be set to <value>false</value> in that case, and should be <value>true</value>
+        /// for new <see cref="Currency">Currencies</see>.</param>
         /// <exception cref="ArgumentException">Thrown when the given <paramref name="ticker"/>
-        /// is an empty string.</exception>
-        public Currency(string ticker, byte decimalPlaces, IImmutableSet<Address>? minters)
+        /// is an empty string, or when <paramref name="totalSupplyTrackable"/> is
+        /// <value>false</value> but <paramref name="maximumSupply"/> is not <value>null</value>,
+        /// or when either the Major or the Minor values of <paramref name="maximumSupply"/> is
+        /// a negative number, or when the given Minor unit for the <paramref name="maximumSupply"/>
+        /// is too big for the given <paramref name="decimalPlaces"/>.</exception>
+        public Currency(
+            string ticker,
+            byte decimalPlaces,
+            (BigInteger Major, BigInteger Minor)? maximumSupply,
```

Comment:
It might be good to have one more overload taking only `maximumSupplyMajor`.

Reply:
Thought about this again, and such overload would confuse code readers, because with no parameter name specified, its order of magnitude could be quite ambiguous, say, whether `1000` means `(1000, 0)` or `(10, 0)` when `decimalPlaces` is 2.

So, now I'm not sure if having such overload is really better.  Let's keep going as it is.

Reply:
👌